taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
+    def default_data(cls, metadata=False, start_year=None):
```

Comment:
Hi Martin,

It seems good to deprecate the `calculator` function, but causing to raise an assertion (as is done in this PR) seems not good, because it breaks existing code in two other repositories. A solution that involves zero additional effort is to merge this PR martinholmer#1 in to your branch. The reasons this is good are as follows:

 - it means dropq and webapp-public will run, instead of raising asserts
 - it doesn't break any tests in this branch
 - it explicitly asserts that the `calculator` function should still work, which it should because it is used in master in dropq.

I don't know is meant by the question: "Does this code produce the same error as T.J. reported above? "

There's no need to run the webapp to demonstrate this issue. One question I asked above was "Are you unable to reproduce this problem?" Let me know if the instructions on reproducing above are unclear.